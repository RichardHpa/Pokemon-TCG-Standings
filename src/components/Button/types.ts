import type { ReactNode, ComponentPropsWithoutRef } from 'react';

export type ButtonProps = {
  children: ReactNode;
  className?: string;
  color?: 'primary' | 'secondary' | 'grey';
  variant?: 'solid' | 'outlined' | 'text';
  size?: 'xs' | 'sm' | 'base' | 'lg' | 'xl';
  disabled?: boolean;
  full?: boolean;
} & ComponentPropsWithoutRef<'button'>;
