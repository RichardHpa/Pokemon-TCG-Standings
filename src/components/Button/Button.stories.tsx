import { Button } from './Button';

import type { Meta, StoryObj } from '@storybook/react';

const meta: Meta<typeof Button> = {
    component: Button,
};
export default meta;

type Story = StoryObj<typeof Button>;

export const Primary: Story = {
    render: () => (
        <div className="flex items-center gap-4">
            <Button>Basic</Button>
            <Button variant="outlined">Basic</Button>
        </div>
    ),
};

export const Secondary: Story = {
    render: () => (
        <div className="flex items-center gap-4">
            <Button color="secondary">Basic</Button>
            <Button color="secondary" variant="outlined">
                Basic
            </Button>
        </div>
    ),
};

export const Size: Story = {
    render: () => (
        <div className="flex items-center gap-4">
            <Button size="xs">extra small</Button>
            <Button size="sm">small</Button>
            <Button>base</Button>
            <Button size="lg">large</Button>
            <Button size="xl">extra large</Button>
        </div>
    ),
};
