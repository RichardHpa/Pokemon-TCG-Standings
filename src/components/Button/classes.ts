export const buttonClasses = {
    base: 'font-medium rounded-lg text-center',
    disabled: 'opacity-50 cursor-not-allowed pointer-events-none',
    full: 'w-full',
    variant: {
        solid: {
            base: 'border',
            primary: {
                light: 'text-white bg-blue-700 hover:bg-blue-800  border-blue-700',
                dark: 'dark:bg-blue-600 dark:hover:bg-blue-700  dark:border-blue-600',
            },
            secondary: {
                light: 'text-gray-900 hover:text-blue-700 bg-white hover:bg-gray-100 border-gray-200  border-white',
                dark: 'dark:text-gray-400 dark:hover:text-white dark:bg-gray-800 dark:hover:bg-gray-600 dark:border-gray-600 dark:border-gray-800',
            },
            grey: {
                light: '',
                dark: '',
            },
        },
        outlined: {
            base: 'border',
            primary: {
                light: 'text-blue-700 hover:text-white border-blue-700 hover:bg-blue-800 focus:ring-blue-300',
                dark: 'dark:text-blue-500 dark:hover:text-white  dark:border-blue-500 dark:hover:bg-blue-500 dark:focus:ring-blue-800',
            },
            secondary: {
                light: 'text-gray-900 hover:text-white hover:bg-gray-900 focus:ring-gray-300 dark:border-gray-600',
                dark: 'dark:text-gray-400 dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-800',
            },
            grey: {
                light: '',
                dark: '',
            },
        },
        text: {
            base: 'border border-transparent',
            primary: {
                light: 'text-blue-700 hover:text-white focus:ring-blue-300',
                dark: 'dark:text-blue-500 dark:hover:text-white dark:focus:ring-blue-800',
            },
            secondary: {
                light: 'text-gray-900 hover:text-white focus:ring-gray-300',
                dark: 'dark:text-gray-400 dark:hover:text-white dark:focus:ring-gray-800',
            },
            grey: {
                light: 'text-gray-500  hover:bg-gray-100 ',
                dark: 'dark:text-gray-400 dark:hover:bg-gray-700',
            },
        },
    },
    size: {
        xs: 'px-3 py-2 text-xs',
        sm: 'px-3 py-2 text-sm',
        base: 'px-5 py-2.5 text-sm',
        lg: 'px-5 py-3 text-base',
        xl: 'px-6 py-3.5 text-base',
    },
};
