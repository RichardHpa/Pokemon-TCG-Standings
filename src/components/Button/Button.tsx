import { useCallback } from 'react';

import clsx from 'clsx';

import { buttonClasses } from './classes';

import type { FC, MouseEvent } from 'react';
import type { ButtonProps } from './types';

export const Button: FC<ButtonProps> = ({
    color = 'primary',
    variant = 'solid',
    size = 'base',
    disabled = false,
    full = false,
    children,
    className,
    onClick,
    ...rest
}) => {
    const handleOnClick = useCallback(
        (event: MouseEvent<HTMLButtonElement>) => {
            event.currentTarget.blur();
            if (onClick) {
                onClick(event);
            }
        },
        [onClick]
    );
    return (
        <button
            type="button"
            disabled={disabled}
            className={clsx(
                buttonClasses.base,
                buttonClasses.size[size],
                buttonClasses.variant[variant].base,
                buttonClasses.variant[variant][color].light,
                buttonClasses.variant[variant][color].dark,
                { [buttonClasses.disabled]: disabled },
                { [buttonClasses.full]: full },
                className
            )}
            {...rest}
            onClick={handleOnClick}
        >
            {children}
        </button>
    );
};
