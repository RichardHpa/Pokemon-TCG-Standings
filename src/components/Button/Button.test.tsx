import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { Button } from './Button';

describe('Button', () => {
  test('renders Button', () => {
    render(<Button>Test</Button>);
    const button = screen.getByRole('button', { name: /test/i });
    expect(button).toBeInTheDocument();
  });

  test('renders Button with onClick', async () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>Test</Button>);
    userEvent.click(screen.getByRole('button'));
    await waitFor(() => expect(onClick).toHaveBeenCalled());
    expect(onClick).toHaveBeenCalledTimes(1);
  });

  test('renders Button with disabled', () => {
    render(<Button disabled>Test</Button>);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  test('renders Button with type', () => {
    render(<Button type="submit">Test</Button>);
    expect(screen.getByRole('button')).toHaveAttribute('type', 'submit');
  });

  test.each([
    ['primary', 'bg-blue-700'],
    ['secondary', 'bg-white'],
  ] as const)('renders Button with color %s', (color, expected) => {
    render(<Button color={color}>Test</Button>);
    expect(screen.getByRole('button')).toHaveClass(expected);
  });

  test.each([
    ['solid', 'text-white'],
    ['outlined', 'text-blue-700'],
  ] as const)('renders Button with variant %s', (variant, expected) => {
    render(<Button variant={variant}>Test</Button>);
    expect(screen.getByRole('button')).toHaveClass(expected);
  });

  test('renders Button with full', () => {
    render(<Button full>Test</Button>);
    expect(screen.getByRole('button')).toHaveClass('w-full');
  });

  test('renders Button snapshot', () => {
    const { container } = render(<Button>Snapshot</Button>);
    expect(container).toMatchSnapshot();
  });
});
