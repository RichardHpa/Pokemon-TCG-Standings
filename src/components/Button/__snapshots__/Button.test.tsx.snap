// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > renders Button snapshot 1`] = `
<div>
  <button
    class="font-medium rounded-lg focus:z-10 focus:ring-4 focus:outline-none text-center px-5 py-2.5 text-sm text-white bg-blue-700 hover:bg-blue-800 focus:ring-blue-300 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
    type="button"
  >
    Snapshot
  </button>
</div>
`;
